src/net/websocketsIO.cpp

```diff
@@ -118,9 +118,15 @@ void WebsocketsClient::wsDisconnect(bool immediate)
     {
         return;
     }
-    
-    assert (thread_id == pthread_self());
-    ctx->wsDisconnect(immediate);
+    else if (ctx->wsIsConnected())
```

Comment:
These changes aim to properly control when the socket is closed by server using `libws`

Reply:
Checking the code of both implementations, it seems that this check doesn't have any impact on any of them.
`libws` checks if `mWebSocket` is `NULL` in both `wsIsConnected` and `wsDisconnect` before doing anything.
The same happens with `libwebsockets` and the field `wsi`.

It seems that `mWebSocket` can be different than `NULL` when the connection is closed by the server because its destruction on `LibwsClient::websockCloseCb` is placed inside a `marshallCall`. But this check wouldn't fix that problem. Instead, I think that it would be better to destroy `mWebSocket` before the `marshallCall` .

Also, I think that it's not good to have the assertion inside the `if` because it makes possible to delete `ctx` from a different thread without triggering the assert.

I see another potential problem related to this code (few lines below). The deletion of `ctx` here will do that `wsIsConnected` returns `false` during graceful disconnects. That's not ideal because that connection is still alive until the disconnection finishes, so there could be unexpected callbacks received in the app. Probably we should only delete `ctx` when we detect that the connection is finished (`ctx->wsIsConnected()` returns `false`) or after immediate disconnections, or when this object is recycled calling `wsConnect` again). Or, even better, we could have an `wsClosePrivateCb` that resets `ctx` and forwards the callback to `wsCloseCb`.